server/utils.go

```diff
@@ -424,42 +390,76 @@ func versionToString(vers int) string {
 	return str
 }
 
-// rewriteToken attempts to match the original token
-// against the email, telephone number or login patterns.
-// On success, prepends the token with the corresponding prefix.
-func rewriteToken(orig, countryCode string) string {
-	if orig == "" || prefixedTagRegexp.MatchString(orig) {
-		// It's either empty or already has a prefix e.g. basic:alice.
+// Tag handling
+
+// Take a slice of tags, return a slice of restricted namespace tags contained in the input.
+// Tags to filter, restricted namespaces to filter.
+func filterRestrictedTags(tags []string, namespaces map[string]bool) []string {
+	var out []string
+	if len(namespaces) == 0 {
+		return out
+	}
+
+	for _, s := range tags {
+		parts := prefixedTagRegexp.FindStringSubmatch(s)
+
+		if len(parts) < 2 {
+			continue
+		}
+
+		if namespaces[parts[1]] {
+			out = append(out, s)
+		}
+	}
+
+	return out
+}
+
+// rewriteToken attempts to match the original token against the email, telephone number and optionally login patterns.
+// The tag is expected to be converted to lowercase.
+// On success, it prepends the token with the corresponding prefix. It returns an empty string if the tag is invalid.
+// TODO: better handling of countryCode:
```

Comment:
I think this TODO is no longer relevant: we are already taking the code from the client provided locale and fallback to the config or hardcoded values.

Reply:
done: changed TODO to be about inferring country code from location